pcdsdevices/targets.py

```diff
@@ -0,0 +1,167 @@
+"""
+Module for common target stage stack configurations.
+"""
+import logging
+
+from ophyd.device import Device
+
+from pcdsdevices.epics_motor import _GetMotorClass
+
+logger = logging.getLogger(__name__)
+
+
+def StageStack(mdict, name):
+    """
+    Conveniencefunction for generating a stage stack device. Intended for
+    bundling various motors into a single object. The function takes a
+    dictionary of PVs and/or previously instantiated motor objects and bundles
+    them together. If given a PV, The factory function attempts to determine
+    the appropriate motor class from the given base PV; if this fails then it
+    will attempt to create an EpicsMotor. Axes are given the same name as they
+    are assigned in the provided dictionary. See examples below.
+
+    Parameters
+    ----------
+    mdict : dictionary
+        Dictionary of motor objects and or base PVs.
+    name : str
+        Name for the stack. Used to make a class name. No whitespace.
+
+    Examples
+    --------
+    # Make a classic XYZ stack with two PVs and one motor object
+    d = {'x': 'TST:MMS:01', 'y': 'TST:MMS:02', 'z': z_motor}
+    xyz = StageStack(d, 'my_xyz')
+    """
+    cpts = {}
+    for mname, mitem in mdict.items():
+        # Check if this is a PV or motor object
+        if issubclass(type(mitem), Device):  # Motor object
+            cpts[mname] = mitem
+        elif isinstance(mitem, (str)):  # PV
+            mcls = _GetMotorClass(mitem)
+            cpt = mcls(prefix=mitem, name=mname)
+            cpts[mname] = cpt
+        else:  # Something is wrong
+            logger.warning("Unrecognized input {}. "
+                           "Skipping axis {}.".format(mitem, mname))
+    cls_name = name + '_StageStack'
+    cls = type(cls_name, (object,), cpts)
+
+    dev = cls()
+
+    return dev
+
+
+class GridAxis():
+    """
+    Class for axes that move in regularly spaced intervals.
+    """
+    def __init__(self, stage, spacing):
+        assert isinstance(spacing, float), "Specify a float target spacing"
+        self.stage = stage
+        self.spacing = spacing
+
+    def advance(self, nspaces, direction, wait=False):
+        """
+        Move the grid axis in the specified direction by (n_spacings * spacing)
+        """
+        assert direction in [1, -1], "Direction must be 1 or -1"
+        assert nspaces >= 1, "n_targets must be >= 1"
```

Comment:
Note that there are situations where Python skips assert statements, though I don't think we're running into this in our current python envs:
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement